pkg/clusterregistry/testing/testserver.go

```diff
@@ -0,0 +1,159 @@
+	// kernel would give us a port which is free for sure), so we need this
+	// workaround.
+	for retry := 0; retry < 5 && !t.Failed(); retry++ {
+		config, td, err := startTestServer(t)
+		if err == nil {
+			return config, td
+		}
+		if err != nil && !strings.Contains(err.Error(), "bind") {
+			break
+		}
+		t.Logf("Bind error, retrying...")
+	}
+
+	t.Fatalf("Failed to launch server")
+	return nil, nil
+}
+
+// startTestServer starts a etcd server and cluster registry API server. Returns
+// a REST client config and a tear-down function.
+//
+// Note: we return a tear-down func instead of a stop channel because the latter
+//       will leak temporary files that becaues Golang testing's call to os.Exit
+//       will not give a stop channel go routine enough time to remove
+//       temporary files.
+func startTestServer(t *testing.T) (result *restclient.Config, tearDownForCaller TearDownFunc, err error) {
+	var tmpDir string
+	var etcdServer *etcdtesting.EtcdTestServer
+	stopCh := make(chan struct{})
+	tearDown := func() {
+		close(stopCh)
+		if etcdServer != nil {
+			etcdServer.Terminate(t)
+		}
+		if len(tmpDir) != 0 {
+			os.RemoveAll(tmpDir)
+		}
+	}
+	defer func() {
+		if tearDownForCaller == nil {
+			tearDown()
+		}
+	}()
+
+	t.Logf("Starting etcd...")
+	etcdServer, storageConfig := etcdtesting.NewUnsecuredEtcd3TestClientServer(t, install.Scheme)
+
+	tmpDir, err = ioutil.TempDir("", "cluster-registry-clusterregistry-apiserver")
+	if err != nil {
+		return nil, nil, fmt.Errorf("Failed to create temp dir: %v", err)
+	}
+	port, err := freePort()
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to find a free port: %v", err)
+	}
+	s := options.NewServerRunOptions()
+	s.SecureServing.BindPort = port
+	s.SecureServing.ServerCert.CertDirectory = tmpDir
+	s.Etcd.StorageConfig = *storageConfig
+	s.Etcd.DefaultStorageMediaType = "application/json"
+
+	t.Logf("Starting cluster registry...")
+	server, err := clusterregistry.CreateServer(s)
+	if err != nil {
+		return nil, nil, fmt.Errorf("Failed to create server: %v", err)
+	}
+	runErrCh := make(chan error, 1)
+	go func(stopCh <-chan struct{}) {
+		if err := server.PrepareRun().Run(stopCh); err != nil {
+			t.Logf("cluster registry exited uncleanly: %v", err)
+			runErrCh <- err
+		}
+	}(stopCh)
+
+	t.Logf("Waiting for /healthz to be OK...")
+	client, err := kubernetes.NewForConfig(server.LoopbackClientConfig)
+	if err != nil {
+		return nil, nil, fmt.Errorf("Failed to create a client: %v", err)
+	}
+	err = wait.Poll(100*time.Millisecond, 30*time.Second, func() (bool, error) {
+		select {
+		case err := <-runErrCh:
+			return false, err
+		default:
+		}
+
+		result := client.CoreV1().RESTClient().Get().AbsPath("/healthz").Do()
+		status := 0
+		result.StatusCode(&status)
+		if status == 200 {
+			return true, nil
+		}
+		return false, nil
+	})
+	if err != nil {
+		return nil, nil, fmt.Errorf("Failed to wait for /healthz to return OK: %v", err)
```

Comment:
I am not entirely sure if this error message is appropriate. It should be something like - "failed waiting for /healthz to ...", no?

Reply:
Good point. I've updated it.